File: clojure/robot-name/robot.clj
(ns robot)

(def ^:private alphabet
  (let [A 65
        Z 90]
    (map char (range A (inc Z)))))

(defn- generate-id []
  (format "%s%s%03d"
          (rand-nth alphabet)
          (rand-nth alphabet)
          (rand-int 1000)))

(defrecord Robot [name])

(defn robot []
  (Robot. (atom (generate-id))))

(defn robot-name [robot]
  @(:name robot))

(defn reset-name [robot]
  (reset! (:name robot) (generate-id)))

File: ruby/robot-name/robot.rb
class Robot
  LETTERS = ('a'..'z').to_a.freeze
  private_constant :LETTERS

  attr_reader :name

  def initialize
    @name = '%s%s%03d' % [LETTERS.sample, LETTERS.sample, rand(100)]
  end

  def reset
    initialize
  end
end
